### PR TITLE
Load page templates via theme.json abstractions

### PR DESCRIPTION
**Pull request by oandregal**
This PR is a follow up to https://github.com/WordPress/gutenberg/pull/27778 (which introduced the `pageTemplates` field in theme.json) and https://github.com/WordPress/gutenberg/pull/28110 https://github.com/WordPress/gutenberg/pull/28533 that refactored it to accommodate the new needs for this file https://github.com/WordPress/gutenberg/issues/28163

The idea is to mature the APIs we use to read data from theme.json, as to:

1. Minimize costly operations: read and parse JSON file, read from the database, etc.
2. Provide abstractions over how to access the underlying data (files in the case of core and theme data / CustomPostType in the case of user data).
3. Centralize logic so the hard parts are easier (things like merging core/theme/user data, translations, rearranging keys, etc).

An initial list of steps we need to check when adding new fields to `theme.json`:

- Add the new field to the theme.json schema.
- Does the field needs any special validation/escaping?
- Does this field needs to merge values from core, theme, and user origins?
- Add unit tests to cover the use case.
- Does this field needs to be translated?

While the role of the `WP_Theme_JSON` class in this space seems well established, the role of `WP_Theme_JSON_Resolver` needs to be accommodated to the new use cases we want to support beyond settings & styles. This PR is the first step to do that and uses `pageTemplates` as an example.

Proposed API for consumers:

```php
$resolver = WP_Theme_JSON_Resolver();
$resolver->get_theme_data()->get_page_templates();
```

## How to test

- Use and activate the TT1-blocks (latest `master` branch). It defines one page template: page-home whose text is "Page without title".
- Load a page and make sure it appears in the template selector (page sidebar > page attributes):
Captura de ecrã de 2021-02-04 14-21-55
- Go to the `experimental-theme.json` of the TT1-blocks theme and update the text to be "My home page".
- Reload the editor and verify the new text is present in the template selector.

## Follow-ups

This PR refactors code, it doesn't change any behavior. A follow up to this that adds new behavior is a PR that enables the page template title to be translated.

I think `WP_Theme_JSON_Resolver` needs more maturation. I'd like to create small and focused follow-up PRs to improve it. It'll allow us to move quickly by making reviews easier.

### CODE REVIEW

File: lib/class-wp-theme-json-resolver.php [hunk 1]
```diff
@@ -258,27 +258,39 @@ private static function get_core_origin() {
...
 	 */
-	private function get_theme_origin( $theme_support_data = array() ) {
-		$theme_json_data = self::get_from_file( locate_template( 'experimental-theme.json' ) );
-		self::translate_presets( $theme_json_data, wp_get_theme()->get( 'TextDomain' ) );
+	public function get_theme_data( $theme_support_data = array() ) {
+		if ( null === $this->theme ) {
+			$theme_json_data = self::get_from_file( locate_template( 'experimental-theme.json' ) );
+			self::translate_presets( $theme_json_data, wp_get_theme()->get( 'TextDomain' ) );
+			$this->theme = new WP_Theme_JSON( $theme_json_data );
+		}
```
Comment (Addison-Stavlo) [inline lib/class-wp-theme-json-resolver.php:280]: Regarding follow up improvements that were mentioned:  When we instantiate the class, it seems `$this->theme` is always `null`.  So if we are calling a `new WP_Theme_JSON_Resolver()` in one place for page templates, and another place for template parts (or anything else).. does that mean we are getting and translating the file separately every time?  Should `theme` itself be static?  Or maybe I am missing something already in place that prevents this from happening more than necessary?

**Review by Addison-Stavlo — APPROVED** [review]
This looks great and testing works well.

I hope you don't mind - I pushed a commit of `composer format` to make the php tests happy and hopefully pass.

File: lib/class-wp-theme-json-resolver.php (hunk 1 above)
Comment (nosolosw) [reply lib/class-wp-theme-json-resolver.php]: Yes! I pushed https://github.com/WordPress/gutenberg/pull/28700/commits/6ed01a9a91b581a58a10d894f9627a39c0c1bfb5 to fix it (I forgot about this, tbh).

**Comment by nosolosw** [comment]
> I hope you don't mind - I pushed a commit of composer format to make the php tests happy and hopefully pass.

Thank you 🙂

File: lib/full-site-editing/page-templates.php [hunk 2]
```diff
@@ -17,17 +17,12 @@ function gutenberg_load_block_page_templates( $templates, $theme, $post ) {
...
-	if ( ! file_exists( $config_file ) ) {
-		return $templates;
-	}
-	$data           = json_decode(
-		file_get_contents( $config_file ),
-		true
-	);
+
+	$resolver       = new WP_Theme_JSON_Resolver();
+	$data           = $resolver->get_theme_data()->get_page_templates();
```
Comment (Addison-Stavlo) [inline lib/full-site-editing/page-templates.php:22]: Circling back to this.  Now that `theme` is static.  Do we need to call `new` here?  Or would something like:
`WP_Theme_JSON_Resolver::get_theme_data()->get_page_templates();` work?

Comment (nosolosw) [reply lib/full-site-editing/page-templates.php:22]: Here's the follow-up that bundles non-behavioral changes: https://github.com/WordPress/gutenberg/pull/28855 (it also makes `get_theme_data` static).

**Pull request merged**